### PR TITLE
Workbench

### PR DESCRIPTION
**Pull request by jakeaturner**

### CODE REVIEW

**Comment by jakeaturner** [comment]
NEW ENV VARIABLES:
CXONE_API_USERNAME=LibreBot
AWS_SSM_LIB_TOKEN_PAIR_PATH=
AWS_SSM_ACCESS_KEY_ID=
AWS_SSM_SECRET_KEY=
AWS_SSM_REGION=

File: server/util/helpers.js [hunk 1]
```diff
@@ -287,4 +287,24 @@ export function getPaginationOffset(page, offsetMultiplier = 25) {
...
 
   return offset;
+}
+
+/**
+ * Breaks up a url into the subdomain and path
+ * @param {string} url
+ * @returns {string[]}
+ */
+export function parseURL(url) {
```
Comment (ethanaturner) [inline server/util/helpers.js:297]: `parseLibreTextsURL`

File: server/util/projectutils.js [hunk 2]
```diff
@@ -651,3 +654,64 @@ export async function checkIfBookLinkedToProject(libreLibrary, libreCoverID) {
     return true;
   }
 }
+
+export async function updateTeamWorkbenchPermissions(projectID, subdomain, coverID) {
+  try {
+    if (!projectID) {
+      throw new Error("Invalid projectID passed to addTeamToWorkbench");
```
Comment (ethanaturner) [inline server/util/projectutils.js:661]: "Invalid projectID passed to updateTeamWorkbenchPermissions"

File: server/util/projectutils.js [hunk 3]
```diff
@@ -651,3 +654,64 @@ export async function checkIfBookLinkedToProject(libreLibrary, libreCoverID) {
...
+    const libUsers = await getLibUsers(subdomain);
+    const foundUsers = libUsers.filter((u) =>
+      centralIDs.includes(u.username.toString())
+    );
+
+    const body = MindTouch.Templates.PUT_TeamAsContributors(
+      foundUsers.map((u) => u.id)
+    );
+
+    console.log(body)
```
Comment (ethanaturner) [inline server/util/projectutils.js:691]: Remove

File: client/src/utils/projectHelpers.ts [hunk 4]
```diff
@@ -21,3 +21,7 @@ export async function pinProject(projectID: string) {
     return false;
   }
 }
+
+export function buildWorbenchURL(libreLibrary: string, libreCoverID: string) {
```
Comment (ethanaturner) [inline client/src/utils/projectHelpers.ts:25]: `buildWorkbenchURL`

File: server/util/CXOne/CXOneTemplates.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  </security>`,
+  PUT_TeamAsContributors: (userIDs: string[]) =>
+    `<security>
+    <permissions.page>
+      <restriction>Semi-Private</restriction>
+    </permissions.page>
+    <grants>
+    <grant>
+      <permissions><role>Manager</role></permissions>
+      <user id="15"></user>
```
Comment (ethanaturner) [inline server/util/CXOne/CXOneTemplates.ts:29]: Use dynamic LibreBot id

File: server/util/projectutils.js [hunk 6]
```diff
@@ -651,3 +654,64 @@ export async function checkIfBookLinkedToProject(libreLibrary, libreCoverID) {
...
+      query: {
+        cascade: "delta",
+      },
+    });
+
+    if (!permsRes.ok) {
+      throw new Error(
+        `Error updating permissions for Workbench book: "${title}"`
+      );
+    }
```
Comment (ethanaturner) [inline server/util/projectutils.js:712]: return `true` for consistent return type

**Comment by jakeaturner** [comment]
ENV config updated

File: client/src/components/projects/ProjectView.jsx [hunk 7]
```diff
@@ -2666,28 +2671,40 @@ const ProjectView = (props) => {
...
+                      <Form.Input
+                        name='projectURL'
+                        type='url'
+                        placeholder='Enter project URL...'
+                        onChange={(e) => setProjURL(e.target.value)}
+                        value={projURL}
+                        id='projectURL'
+                      />
+                      </Form.Field>
+                      <Button color='blue' onClick={() => setShowCreateWorkbenchModal(true)} className='!mb-4'><Icon name='plus'/>New Book</Button>
```
Comment (ethanaturner) [inline client/src/components/projects/ProjectView.jsx:2699]: "Create Book"

File: client/src/components/projects/ProjectView.jsx [hunk 8]
```diff
@@ -2666,28 +2671,40 @@ const ProjectView = (props) => {
...
+                      </>
+                  )
+                }
+                {
+                  project.didCreateWorkbench && (
+                    <Label basic color='green' className='!mb-4 cursor-pointer' onClick={() => window.open(buildWorbenchURL(project.libreLibrary, project.libreCoverID))}>Workbench created <Icon name='external' className='!ml-1'/></Label>
```
Comment (ethanaturner) [inline client/src/components/projects/ProjectView.jsx:2705]: "Book created"

File: client/src/components/projects/CreateWorkbenchModal.tsx [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+      onSuccess();
+    } catch (err) {
+      handleGlobalError(err);
+    } finally {
+      setLoading(false);
+    }
+  }
+  return (
+    <Modal size="fullscreen" {...rest}>
+      <Modal.Header>Create Workbench</Modal.Header>
```
Comment (ethanaturner) [inline client/src/components/projects/CreateWorkbenchModal.tsx:74]: "Create Book"

File: client/src/components/projects/CreateWorkbenchModal.tsx [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+                  Use Dev
+                </p>
+              </>
+            )}
+          </div>
+          <div className="mt-4">
+            <CtlTextInput
+              control={control}
+              name="title"
+              label="Workbench Title"
```
Comment (ethanaturner) [inline client/src/components/projects/CreateWorkbenchModal.tsx:125]: "Book Title"

File: client/src/components/projects/CreateWorkbenchModal.tsx [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+              placeholder="Enter Workbench Title"
```
Comment (ethanaturner) [inline client/src/components/projects/CreateWorkbenchModal.tsx:126]: "Enter Book Title"

File: client/src/components/projects/CreateWorkbenchModal.tsx [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+              required
+              rules={required}
+            />
+          </div>
+          <p>
+            <strong>CAUTION:</strong> Library cannot be changed after Workbench
```
Comment (ethanaturner) [inline client/src/components/projects/CreateWorkbenchModal.tsx:132]: ...after "Book" is created...

File: server/api/books.ts [hunk 13]
```diff
@@ -1345,6 +1361,158 @@ async function getCatalogFilterOptions(_req: Request, res: Response) {
   }
 }
 
+/**
+ * Creates a new book with default features in a user's sandbox.
```
Comment (ethanaturner) [inline server/api/books.ts:1365]: ...in a library Workbench area...

File: server/api/books.ts [hunk 14]
```diff
@@ -1345,6 +1361,158 @@ async function getCatalogFilterOptions(_req: Request, res: Response) {
...
+ */
+async function createBook(
+  req: ZodReqWithUser<z.infer<typeof createBookSchema>>,
+  res: Response
+) {
+  try {
+    const { library, title, projectID } = req.body;
+    const { uuid: userID } = req.user.decoded;
+
+    const user = await User.findOne({ uuid: userID }).orFail();
```
Comment (ethanaturner) [inline server/api/books.ts:1378]: return 404 if not found

File: server/api/books.ts [hunk 15]
```diff
@@ -1345,6 +1361,158 @@ async function getCatalogFilterOptions(_req: Request, res: Response) {
...
+    const project = await Project.findOne({ projectID }).orFail();
```
Comment (ethanaturner) [inline server/api/books.ts:1379]: return 404 if not found

File: server/api/books.ts [hunk 16]
```diff
@@ -1345,6 +1361,158 @@ async function getCatalogFilterOptions(_req: Request, res: Response) {
...
+
+    const subdomain = getSubdomainFromLibrary(library);
+    if(!subdomain) {
+      throw new Error("Invalid library");
+    }
+
```
Comment (ethanaturner) [inline server/api/books.ts:1385]: Use `ProjectsAPI.checkProjectMemberPermission()` to verify permission first

File: server/util/librariesclient.ts [hunk 17]
```diff
@@ -0,0 +1,472 @@
...
+      if (path) {
+        const isNumber = !isNaN(Number(path));
+        const queryIsFirst = api?.includes("?") ? false : true;
+        const url = `https://${subdomain}.libretexts.org/@api/deki/users/${
+          isNumber ? "" : "="
+        }${encodeURIComponent(encodeURIComponent(path))}/${api ? api : ""}${_parseQuery(
+          query,
+          queryIsFirst
+        )}`;
+        console.log(url)
```
Comment (ethanaturner) [inline server/util/librariesclient.ts:165]: remove

File: server/util/librariesclient.ts [hunk 18]
```diff
@@ -0,0 +1,472 @@
...
+ * Returns all users of a library.
+ *
+ * @param subdomain - The subdomain of the library
+ * @returns {Promise<CXOneUser[]>} - The users of the library
+ */
+export async function getLibUsers(subdomain: string): Promise<CXOneUser[]> {
+  try {
+    const res = await CXOneFetch({
+      scope: "users",
+      subdomain,
```
Comment (ethanaturner) [inline server/util/librariesclient.ts:367]: You will need to set `?limit=all` or implement pagination.

File: server/util/librariesclient.ts [hunk 19]
```diff
@@ -0,0 +1,472 @@
...
+
+export const getSubdomainFromLibrary = (library: string): string | null => {
+  // TODO: get full list
+  if (library === "chem") return "chem";
+  if (library === "phys") return "phys";
+  if (library === "bio") return "bio";
+  if (library === "eng") return "eng";
+  if (library === "math") return "math";
+  if (library === "stats") return "stats";
+  if (library === "dev") return "dev";
```
Comment (ethanaturner) [inline server/util/librariesclient.ts:470]: Finish, but it seems like you're returning the same value?

**Review by ethanaturner — CHANGES_REQUESTED** [review]
Looking good so far

**Review by ethanaturner — APPROVED** [review]
_ET_. 12/30/23.

**Pull request merged**